### PR TITLE
refactor core to create cluster_slots_service module

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem
The `core` crate can be split up to facilitate taking dependencies on modules without ending up with circular dependencies. Cluster slots modules can be separated from core.

#### Summary of Changes
Separate cluster slot modules into `cluster-slots` crate.

Fixes #

### CODE REVIEW

File: cluster-slots/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+name = "solana-cluster-slots"
+description = "Solana Cluster Slots"
+documentation = "https://docs.rs/solana-cluster-slots"
+version = { workspace = true }
+authors = { workspace = true }
+repository = { workspace = true }
+homepage = { workspace = true }
+license = { workspace = true }
+edition = { workspace = true }
+publish = false
```
Comment (jbiseda) [inline cluster-slots/Cargo.toml:11]: Any thoughts on the name of the crate? I'll need to reserve the name for publishing.

**Comment by mvines** [comment]
> The `core` crate can be split up to facilitate taking dependencies on modules without ending up with circular dependencies. Cluster slots modules can be separated from core.

Sorry just so I better understand, what's the actual motivation for this PR? Just because cluster slots modules **can** be separated from core doesn't necessarily mean it **should** be.

**Comment by jbiseda** [comment]
> Sorry just so I better understand, what's the actual motivation for this PR? Just because cluster slots modules **can** be separated from core doesn't necessarily mean it **should** be.

My initial goal was to separate repair modules into a new crate, this is preliminary work. New crates would be:
- cluster slots
- consensus
- repair

Initial outline of the changes in: split core POC #31667

**Comment by mvines** [comment]
Thanks. Is this actually the minimum set that'd be required by a future consensus and repair crate? Specifically why does `ClusterSlotsService` move? 

> Initial outline of the changes in: split core POC https://github.com/solana-labs/solana/pull/31667

What about starting by refactoring it all within `core/`? So separate subdirectories like `core/src/{cluster-slots,consensus,
repair}`? This provides close to the same level of code isolation without the upfront overhead of making new crates. Also if we don't like how the refactoring feels, it's much easier to fold the code back into `core/src`. Once `core/` internally is neat and tidy, then we might consider if truly creating new crates for parts of core is worth the overhead

**Comment by jbiseda** [comment]
> What about starting by refactoring it all within `core/`? So separate subdirectories like `core/src/{cluster-slots,consensus, repair}`? This provides close to the same level of code isolation without the upfront overhead of making new crates. Also if we don't like how the refactoring feels, it's much easier to fold the code back into `core/src`. Once `core/` internally is neat and tidy, then we might consider if truly creating new crates for parts of core is worth the overhead

Good idea, I'll update this to start the refactoring within the core crate.

**Comment by CriesofCarrots** [comment]
> My initial goal was to separate repair modules into a new crate

This subdirectory reorg looks okay to me, but I'm curious, what is the original motivation for repair modules in a new crate? Is there another downstream user forthcoming?

**Comment by jbiseda** [comment]
> > My initial goal was to separate repair modules into a new crate
> 
> This subdirectory reorg looks okay to me, but I'm curious, what is the original motivation for repair modules in a new crate? Is there another downstream user forthcoming?

The use case I'm interested in is leveraging these modules in test code without requiring `core` as a dependency.

**Review by CriesofCarrots — APPROVED** [review]

**Pull request merged**